anbox: fixes and upgrades

samueldr opened the pull request:
###### Motivation for this change

Make anbox work.

Some changes, mainly geared towards making it work.

Technically some of the changes could be considered breaking... But anbox *doesn't work at all* currently. So it can't be more breaking than that.

The AArch64 image *from anbox* doesn't work on my system. Though the *postmarketOS* image works fine.

<details>
<summary>Outdated info about kernel config</summary>

Newer kernels will require this configuration to be set:

```nix
{
boot.kernelPatches = [
  {
    name = "anbox-options";
    patch = null;
    extraConfig = ''
      ASHMEM y
      ANDROID y
      ANDROID_BINDER_IPC y
      ANDROID_BINDERFS y
    '';
  }
];
}
```

I do not know if we would prefer setting these as defaults. They're not modules, so it changes the way the kernel acts I guess.

</details>

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Tested with:

 - Pinebook Pro, kernel 5.11, with kernel configuration snippet added


###### To be done

 - [x] Test on x86_64
 - ~~Test on kernel with compatible anbox module (e.g. 5.4)~~
 - ~~Validate how it acts on 5.5+ when the kernel is configured "incorrectly"~~
 - [x] Add image edition facilities (to allow end-users to e.g. add F-Droid or a rooting solution...)
 - [x] Look at making the generated desktop files more robust ([hint](https://github.com/anbox/anbox/blob/9de4e87cdd05135e1c71e6eadb68bf82719cebdf/src/anbox/application/launcher_storage.cpp#L72-L74))
 - ~~[Adopt `linux-zen` patches](https://github.com/NixOS/nixpkgs/pull/102341#discussion_r554550133)?~~ No [they reverted them](https://github.com/NixOS/nixpkgs/pull/102341#issuecomment-792173609)
 - ~~Add "OSS" [android-tools](https://github.com/nmeum/android-tools)~~ #124992 already did it... I feel silly.
 - [x] Use those android tools for the anbox test
 - [ ] Give escape hatches for "expert mode use" through internal options; `useAnboxModules` and `useBinderFS` options that defaults according to the current semantics of looking at the kernel version. (Mobile NixOS, and devices using vendor kernels may not fit the requirements, but will actually work.)

* * *

**TO ANYONE TESTING**: Can you please include these informations:

 - Kernel package used (e.g. linuxPackages_latest)
 - How exactly you made `ashmem` and `binderfs` available

mvnetbiz committed:
linux: CONFIG_ASHMEM=y, CONFIG_ANDROID=y

mvnetbiz committed:
nixos/anbox: Cleanups from: use mainline drivers when available

Co-authored-by: Samuel Dionne-Riel <samuel@dionne-riel.com>

mvnetbiz committed:
nixos/tests/anbox: init

samueldr committed:
lxc: Add patch required to unbreak anbox

See https://github.com/lxc/lxc/pull/3844

samueldr committed:
anbox: 2020-11-29 -> 2021-05-26

samueldr committed:
anbox: Fix compatibility with LXC 4

This patch is used actively by postmarketOS, which in turn *just* uses
Alpine Linux packages for anbox.

samueldr committed:
anbox: Be ten times more patient when launching

This helps on slower devices, at the cost of making the command
"resolve" less quickly.

It replaces commands "failing because things weren'd ready", but things
actually still succeeding by commands *possibly* failing in 10× more
time.

samueldr committed:
nixos/anbox: Use anbox kernel modules only on compatible kernels.

Doing so allows using a properly configured new enough kernel with
Anbox.

samueldr committed:
nixos/anbox: Mark anbox0 as unmanaged with NetworkManager

samueldr committed:
nixos/anbox: Manage mounting /dev/binderfs

samueldr committed:
nixos/anbox: Validate kernel config for newer kernels

This uses the wrong assumption that with 5.4 and prior users will use
the anbox module. Though that assumption is made true by the current
NixOS configuration.

samueldr committed:
nixos/anbox: Minor cleanup

samueldr committed:
anbox: Remove (thankfully inactive) "su" default

The `anbox-application-manager` helper, which is actually broken as it's
not installed as +x, starts "su" by default, which might not be
desirable, depending on the end-user's wishes.

samueldr committed:
anbox: Fix and reclaim anbox-application-manager

It now *strictly* is a shortcut to launch the application manager.

samueldr committed:
anbox: Add samueldr to maintainers

samueldr committed:
anbox-postmarketos-image: init at 7.1.2_r39-anbox

edwtjo commented:
Good job (from a cursory inspection)! I will try these out during the weekend.

cidkidnix commented on pkgs/os-specific/linux/anbox/postmarketos-image.nix:
```diff
@@ -0,0 +1,19 @@
+      sha256 = "0dx8mhfcjbkak982zfh65bvy35slz5jk31yl4ara50ryrxsp32nx";
+    };
+    x86_64-linux = fetchurl {
+      url = imgroot + "android-7.1.2_r39-anbox_x86_64-userdebug.img";
```
you're missing a `/` here before the image name

samueldr replied:
Thanks!

cidkidnix commented:
After the fix to the postmarketos image everything works well for me

cidkidnix commented:
![1622820191_screenshot](https://user-images.githubusercontent.com/67574902/120825435-ff6daa80-c51e-11eb-95dd-9de1df9bec69.png)
Everything works well but audio is pretty badly delayed

samueldr commented:
Ah, note that even on another attempt a year back `anbox session-manager` never worked for me. I was using the same ocmmand `anbox-application-manager` aliases to.

Launching the first anbox app from a fresh boot/session *may* take some time, as it has to start the container too. Anbox assumes on a modern system it takes "up to 10 seconds". The assumption was changed to 100 seconds, similar to what postmarketOS does.

samueldr commented:
@cidkidnix can you also tell me the kernel version, and how you made `ashmem` and `binderfs` available? For statistical purposes :).

cidkidnix commented:
```{ config, pkgs, lib, ... }:
{
  config = lib.mkIf config.virtualisation.anbox.enable {
    boot.kernelPatches = [
      {
        name = "ashmem-binder";
        patch = null;
        extraConfig = ''
          ASHMEM y
          ANDROID y
          ANDROID_BINDER_IPC y
          ANDROID_BINDERFS y
          ANDROID_BINDER_DEVICES binder,hwbinder,vndbinder
        '';
      }
    ];
  };
}
```
I patched my nixpkgs with this PR, My kernel version is 5.11.21, and the above code snippet is what I have imported as a module to enable the kernel configs

cidkidnix commented:
I'll also be able to test on my pinephone sometime soon:tm:

samueldr commented:
Looks like the patch for audio requires a fresh image build. So that wouldn't help AFAICT.

It would be nice to see @danielfullmer's robotnix build the Anbox images, so that a more trustful environment can be built to run Anbox.

samueldr committed:
nixos/anbox: Allow applying changes to the image

samueldr commented:
Here's an example for customizing the image.

```nix
{ pkgs, ... }:
{
  virtualisation.anbox = {
    enable = true;
    image = pkgs.anbox-postmarketos-image;

    # https://gitlab.com/postmarketOS/pmaports/-/blob/bf6ad7a78c5506eb5ad9089e87f9c1cf7e8cd1f8/main/anbox-image/APKBUILD#L39-63
    imageModifications =
      let
        fdroid_apk = pkgs.fetchurl {
          url = "https://f-droid.org/repo/org.fdroid.fdroid_1013000.apk";
          sha256 = "1n5zcxsfn42b3i067pnkjy3xf5ljs0fj9h01v69xfincr84mlh0y";
        };
        fdroidpriv_apk = pkgs.fetchurl {
          url = "https://f-droid.org/repo/org.fdroid.fdroid.privileged_2120.apk";
          sha256 = "1axa72vfd8qq2dyk7d171vpwb6rf5ps59zrchhhgqmrqmpv88gww";
        };
      in
    ''
      (PS4=" $ "; set -x

        echo "Disabling su"
        rm -v system/xbin/su

        echo "Installing FDroid"
        mkdir system/app/FDroid
        mkdir system/priv-app/FDroid

        # install -Dm 0644 doesn't work as it copies the symlink, hence cp
        cp --dereference "${fdroid_apk}"     system/app/FDroid/${fdroid_apk.name}
        cp --dereference "${fdroidpriv_apk}" system/priv-app/FDroid/${fdroidpriv_apk.name}

        chmod 0644 system/app/FDroid/$_fdroid
        chmod 0644 system/priv-app/FDroid/$_fdroidpriv
      )
    '';
  };
}
```

cidkidnix commented:
Could we not make a script that runs once after anbox is run to `adb install` applications?

samueldr committed:
anbox: Ensure .desktop files use a correct `anbox`

This fixes two issues

  - store path is hardcoded in desktop files
  - `.anbox-wrapped` is used in desktop files

samueldr commented:
> Could we not make a script that runs once after anbox is run to adb install applications?

Running once is... harder than it sounds. And doesn't handle removing apps.

The best bet actually would be to force the desired apps at the system level just like F-Droid here. Though I wonder how things act once a system app is removed. What happens with its data?

Still, it's out of scope for these changes. The foundations are now here to work on more complex behaviours if desired.

I do not intend to work on fully nix-ifying the Anbox system config, it'd be hard for not enough benefits.

I think installing F-Droid here to allow you to *then* install apps you want, and have F-Droid managing updates is a good enough solution for what hopefully is a temporary problem. We don't want to rely on Android apps!

cidkidnix commented:
A massive change will probably be needed when erfan's A10 anbox stuff for halium gets ready to use, but we can deal with that when it happens https://github.com/Anbox-halium

cidkidnix commented:
To add on, when a system app is install all data *should* be removed, although that isn't always the case

mvnetbiz commented:
>  Test on x86_64

Works with setting `kernelPatches` and after running manually `mount -t binder binder /dev/binderfs`

> Test on kernel with compatible anbox module (e.g. 5.4)

application-manager starts on 5.4

> Validate how it acts on 5.5+ when the kernel is configured "incorrectly"

On 5.10 without kernel CONFIG_ANDROID options, anbox-application-manager window stays open for a minute, then exits.
```
Jun 05 03:36:54 nixos anbox[1428]: Failed to start as either binder or ashmem kernel drivers are not loaded
Jun 05 03:36:54 nixos dbus-daemon[1111]: [session uid=1000 pid=1111] Activated service 'org.anbox' failed: Process org.anbox exited with status 1
```

mvnetbiz commented:
For some reason it is not mounting /dev/binderfs for me. I'm trying to figure out, but you could make it a mount unit required by the container-manager like here.  https://github.com/NixOS/nixpkgs/pull/102341/files#diff-b05067fd6d27b2c89be023bde744ecaab904222a09a136d004f9e5c5148fe966R81-R87

samueldr commented:
@mvnetbiz

Please tell me your:

 -  Kernel version

Then questions:

 - Are you using the anbox modules or are you using the amended kernel configuration?
 - Did you reboot after rebuilding with the amended kernel configuration?
 - What is the error messages from the failure to mount /dev/binderfs?

samueldr commented:
Note that I don't know how this acts on 5.4 yet. I still need to get a "normal" computer that runs 5.4 nicely to test a bit with the anbox kernel modules.

I assume it's possible `/dev/binderfs` won't mount in that case. And it's possibly fine?

samueldr commented:
> On 5.10 without kernel CONFIG_ANDROID options, anbox-application-manager window stays open for a minute, then exits.

Yes, that's to be expected. Nothing we can do if the kernel options are not enabled. They are features Anbox and Android uses.

mvnetbiz commented:
I've been using on my main computer, but to test this, I'm just using a simple nixos configuration with the qemu-vm output. I am uncommenting 1 of the two sections to test with the built in drivers or the out of tree ones.
```nix
{ config, lib, pkgs, ... }:

{
  imports = [
    ../nixos/modules/installer/cd-dvd/installation-cd-graphical-plasma5.nix
    ../nixos/modules/virtualisation/qemu-vm.nix
  ];

  #boot.kernelPackages = pkgs.linuxPackages_5_4;

  #boot.kernelPatches = [
  #  {
  #    name = "anbox-options";
  #    patch = null;
  #    extraConfig = ''
  #      ASHMEM y
  #      ANDROID y
  #      ANDROID_BINDER_IPC y
  #      ANDROID_BINDERFS y
  #    '';
  #  }
  #];


  programs.adb.enable = true;

  virtualisation.anbox.enable = true;
  virtualisation.memorySize = 4000;
  virtualisation.cores = 2;
}
```

mvnetbiz commented:
> > On 5.10 without kernel CONFIG_ANDROID options, anbox-application-manager window stays open for a minute, then exits.
> 
> Yes, that's to be expected. Nothing we can do if the kernel options are not enabled. They are features Anbox and Android uses.

Right, I was not sure what you meant by
> Validate how it acts on 5.5+ when the kernel is configured "incorrectly"

So I just tested just to make sure it crashes correctly when the options aren't enabled :)

mvnetbiz commented:
Or did you mean what warnings or errors are asserted at evaluation time of nixos config?

samueldr commented:
Ah! Sorry, I thought it was a "support question" and not an answer to my initial questions.

And I meant more about the NixOS assertions. I hoped it wouldn't build without the kernel configuration options being set.

mvnetbiz commented:
I don't know why the `fileSystems` option isn't working. It doesn't even appear to be set.
```
env NIX_PATH='nixpkgs=/home/matt/src/nixpkgs:nixos-config=/home/matt/src/nixpkgs/anbox-test/anbox.nix' nixos-option fileSystems | grep binder
```
nothing.

samueldr commented:
I also wasn't aware of your previous PR. I did search through PRs, but a lot of unrelated PRs were present in the results so I missed the only one relevant PR >.<.

I'll look and see if anything is better or should be used instead. If so will transplant the commits.

mvnetbiz commented:
That's fine, thanks for updating this more, the container image options are awesome!

mvnetbiz commented:
I also added a NixOS test in my PR. It might be a good idea to make it take multiple kernels as a parameter? There were concerns about `adb` license, raised in the comments, but to me it looks like that is free? Also the test seems slightly flakey to me, not sure if that is a memory issue or what, but maybe you have some insight into that.

samueldr commented:
The license issue comes from the fact that the adb package we use comes from the Android SDK, which *it* requires an EULA to be agreed. See the comment I left on the other PR for a solution :).

Note that this also is an issue for aarch64-linux. No fastboot, no adb. The SDK is only produced for x86_64-linux!

samueldr commented:
As for the kernel config... I guess we really can do it like your example, and always build our kernels with the options enabled.

We'd need to look more into the issues those options can cause, if any. Though supposedly Ubuntu has it enabled by default if I understand things right?

mvnetbiz commented:
binderfs is only mounted when anbox is enabled, and I'm not sure if I fully understand, but it looks like previous ashmem vulnerabilities were at least mostly due to corrupting shared memory of other processes, which are at least partially sandboxed here, or not running at all if anbox isn't enabled.

mvnetbiz commented:
Pushed new version with `android-tools`, if you want to pick that. Also about testing multiple kernel versions, I figure the test is more to make sure anbox package keeps working, an older kernel will not likely break this I think.

samueldr commented:
Yeah, if I follow your lead and force the Anbox module to require 5.x kernels (with an escape hatch for android vendor kernels), and always apply the ashmem/binderfs config to the kernel, then there's no need to test more than one kernel version imo.

samueldr committed:
nixos/anbox: Use a systemd mount for binderfs

Cleaner than forcing a "full" filesystem mount.

Co-authored-by: Matt Votava <mvnetbiz@gmail.com>

samueldr committed:
nixos/anbox: Use `--daemon` to squash warning

As noted in https://github.com/NixOS/nixpkgs/pull/102341 this is not
actually running as a forked process. It only tells the process that it
is running "as a daemon, so shut the warning up".

See `daemon_` here

 - https://github.com/anbox/anbox/blob/9de4e87cdd05135e1c71e6eadb68bf82719cebdf/src/anbox/cmds/container_manager.cpp#L38-L79

It is **strictly** used to hide that message.

Co-authored-by: Matt Votava <mvnetbiz@gmail.com>

samueldr committed:
anbox: Fix `WM_CLASS`

SDL defaults to reading the executable from the running app

 - https://github.com/libsdl-org/SDL/blob/64724db0a1ffac7c3332df66a942d3ba27b2dc0b/src/video/x11/SDL_x11video.c#L69-L71

Which in turn gives this from xprop:

```
WM_CLASS(STRING) = ".anbox-wrapped", ".anbox-wrapped"
```

Meaning that any special casing a window manager does will be broken.

This will also fallback correctly on Wayland:

 - https://github.com/libsdl-org/SDL/blob/d956636c85aafc055d37153d52370e9b1c2c5929/src/video/wayland/SDL_waylandvideo.c#L93-L99

cidkidnix commented:
We might want to add a systemd-user service option that autostarts `anbox session-manager`. I'm not sure if this should be in this PR or a separate one.

samueldr commented:
I don't think it's neded. What does this service actually do?

AFAIUI it's not needed, as upstream anbox doesn't seem to rely on it for any behaviour anymore? The `.desktop` files from upstream just start intents as needed. I never ended up using it in my testing.

cidkidnix commented:
Personally my service just runs `anbox session-manager`at login so I know anbox is booted by the time that I get around to actually  testing it, and removes issues where the .desktop file just doesn't start it at all

samueldr commented:
> and removes issues where the .desktop file just doesn't start it at all

Could this be fixed by [this commit](https://github.com/NixOS/nixpkgs/pull/125600/commits/c9efa4fd7c5582a35da0acd392402dd2b5cadafc)? It stops baking a full store path in the desktop files.

I've never seen anbox not starting from a desktop file... But I have not used it before working on this PR.

samueldr committed:
anbox: Add patch to provide window icon

Currently the windows don't provide icons. It's rather inconvenient for
many reasons.

This patch forwards the icons from the Android application to the
window.

Patch origin: https://fjordtek.com/git/Fincer/anbox-install/commit/89226287860cf8ced3d41868c80cfa216b5237af

Rebased on top of the current tip of anbox master

See also: https://github.com/anbox/anbox/issues/1514

samueldr committed:
nixos/tests/anbox: Use postmarketOS anbox image

The AArch64 image from anbox will not start.

samueldr commented:
With this last push, I have verified:

 - [x] the test runs succesfully on AArch64
 - [x] the test runs succesfully on x86_64

In addition a patch to provide icons to the windows was added. It's a relatively minor change, but with a huge impact as it makes Anbox more pleasant to use.

mvnetbiz commented:
Tested again, both images on x86_64 are working.

cidkidnix commented:
I'll test but it seems anbox crashes for me when it's running in the background and restarts itself, temporarily(maybe?) causing my user to lose connection to systemd, started happening shortly after I enabled anbox

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/run-android-apps-in-nixos/14437/2

SuperSandro2000 commented on pkgs/os-specific/linux/anbox/default.nix:
```diff
@@ -85,7 +79,7 @@ stdenv.mkDerivation rec {
   ];
 
-  patchPhase = ''
+  prePatch = ''
```
```suggestion
  postPatch = ''
```

otherwise patches no longer rapply.

asdf8dfafjk commented:
@cidkidnix btw, anbox halium is at github.com/waydroid/anbox-halium

alyssais commented:
The lxc patch seems to no longer be required in latest nixpkgs.

asdf8dfafjk commented:
Also relevant: https://github.com/NixOS/nixpkgs/pull/140200/commits/5fc5a2d84cb566abe8f6125a750612c6c5ddd7db

arximboldi commented:
Sorry if this is a very newbie question, but is there a way to try to this PR without bringing this commit into the channels nor replacing my whole nixpkgs by this -- i.e. just using the relevant packages and options from here? I know how to do that when it's just one package to add to `environemnt.systemPackages` or alike, but this seems a bit more involved.

Atemu commented:
Best just clone nixpkgs, merge this branch and then `nixos-rebuild test -I nixpkgs=/path/to/checkout/`

arximboldi commented:
Aha! That's easy and nice, thank you @Atemu I didn't know I could just do `test -I`

asdf8dfafjk commented:
@arximboldi I don't want to disrespect anybody's work but at this point you should probably try https://github.com/NixOS/nixpkgs/pull/141076 since waydroid is newer than anbox

mzabani commented:
Merged this into a relatively recent 21.11 nixpkgs branch, removed the lxc patch as suggested by @alyssais and it seems to work. Is something other than a suggested change 
by @SuperSandro2000 holding this? Is there something I/we can do to help get this merged?

SuperSandro2000 commented on nixos/modules/virtualisation/anbox.nix:
```diff
@@ -25,6 +29,28 @@ let
+        squashfsTools
+      ];
+    } ''
+      echo "→ Extracting Anbox root image..."
```
```suggestion
      echo "-> Extracting Anbox root image..."
```
we shouldn't use special unicode characters

SuperSandro2000 commented on nixos/modules/virtualisation/anbox.nix:
```diff
@@ -25,6 +29,28 @@ let
+      ${cfg.imageModifications}
+      )
+
+      echo "→ Packing modified Anbox root image..."
```
```suggestion
      echo "-> Packing modified Anbox root image..."
```

SuperSandro2000 commented on pkgs/os-specific/linux/anbox/postmarketos-image.nix:
```diff
@@ -0,0 +1,19 @@
+      url = imgroot + "/android-7.1.2_r39-anbox_x86_64-userdebug.img";
+      sha256 = "16vmiz5al2r19wjpd44nagvz7d901ljxdms8gjp2w4xz1d91vzpm";
+    };
+  }.${stdenv.system}
```
Missing or throw unsupported system to eval on unsupported platforms.

SuperSandro2000 commented:
> by @SuperSandro2000 holding this? Is there something I/we can do to help get this merged?

I can't really help because I don't know kernels to well.

Atemu commented:
The kernel stuff should be dropped anyways. It was necessary to support kernels <5.5 which we don't really need to anymore.

Mindavi commented:
Kernel config was added here: https://github.com/NixOS/nixpkgs/pull/145768

rnhmjoj commented:
Continuation in https://github.com/NixOS/nixpkgs/pull/253146.
